# SexSpecificScaffolds-FUNCTIONS.R
# functions for testing scaffold depth and variant data for sex chromosome signals
# REQUIRES: base R

# make_df.sexbias() a function to build the df.sexbias dataframe
# takes a list of 5 filenames, all generated by Ceres scripts: 
#   faidx: the assembly fasta index (for scaffold lengths)
#   depthF: depth_per_scaffold.tsv from female reads, best to use hardmasked assembly
#   depthM: depth_per_scaffold.tsv from male reads, best to use hardmasked assembly
#   nvariF: nseg_per_scaffold.csv from female reads, count of variant sites
#   nvariM: nseg_per_scaffold.csv from male reads, count of variant sites
# reads in all files, sorts by scaffold length, calculates for each scaffold:
#   - proportion of total assembly length
#   - normalized depths (to length-weighted average depth), female and male
#   - frequency of variant sites, female and male
#   - F/M ratio of normalized depth
#   - F/M ratio of variant frequency 
make_df.sexbias <- function(l.IN_FILES){
  # get scaffolds and lengths from asm fasta index
  FAIDX <- read.table(l.IN_FILES$faidx, 
                      col.names = c('scaffold', 'length', 'x','y','z'), 
                      colClasses = c("character", "integer", "NULL", "NULL", "NULL"))
  # input dataframes
  IN_DEPTHF <- read.table(l.IN_FILES$depthF, header = F, row.names = 1, 
                          col.names = c('scaffold', 'depth'))
  IN_DEPTHM <- read.table(l.IN_FILES$depthM, header = F, row.names = 1, 
                          col.names = c('scaffold', 'depth'))
  IN_NVARIF <- read.csv(l.IN_FILES$nvariF, header = F, row.names = 1, 
                        col.names = c('scaffold','n_variants'))
  IN_NVARIM <- read.csv(l.IN_FILES$nvariM, header = F, row.names = 1, 
                        col.names = c('scaffold','n_variants'))
  # scaffolds ordered by length
  SCAFFOLDS <- FAIDX[order(-FAIDX$length),]$scaffold
  # start output df ordered by scaffold length
  SEXBIAS <- data.frame(scaffold = SCAFFOLDS,
                        length = FAIDX[order(-FAIDX$length),]$length,
                        depthF = IN_DEPTHF[SCAFFOLDS,],
                        depthM = IN_DEPTHM[SCAFFOLDS,],
                        n_varF = IN_NVARIF[SCAFFOLDS,],
                        n_varM = IN_NVARIM[SCAFFOLDS,])
  SEXBIAS[is.na(SEXBIAS)] = 0
  # proportion of total length per scaffold
  SEXBIAS$pr_length <- SEXBIAS$length / sum(SEXBIAS$length)
  # length-weighted normalized depths
  SEXBIAS$depthF_norm <- SEXBIAS$depthF / sum(SEXBIAS$depthF * SEXBIAS$pr_length)
  SEXBIAS$depthM_norm <- SEXBIAS$depthM / sum(SEXBIAS$depthM * SEXBIAS$pr_length)
  # frequency of variant sites
  SEXBIAS$fr_varF <- SEXBIAS$n_varF / SEXBIAS$length
  SEXBIAS$fr_varM <- SEXBIAS$n_varM / SEXBIAS$length
  # sex bias (female / male)
  SEXBIAS$Depth_bias <- SEXBIAS$depthF_norm / SEXBIAS$depthM_norm
  SEXBIAS$FrVar_bias <- SEXBIAS$fr_varF / SEXBIAS$fr_varM
  
  return(SEXBIAS)
  
}

